mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Support for updating the mempool upon a reorg
+
+use std::collections::BTreeSet;
+
+use chainstate::chainstate_interface::ChainstateInterface;
+use common::{
+    chain::{Block, GenBlock, SignedTransaction},
+    primitives::{Id, Idable},
+};
+use logging::log;
+use utils::tap_error_log::LogError;
+use utxo::UtxosStorageRead;
+
+use crate::{get_memory_usage::GetMemoryUsage, pool::Mempool};
+
+/// An error that can happen in mempool on chain reorg
+#[derive(Debug, thiserror::Error, PartialEq, Eq)]
+pub enum ReorgError {
+    #[error(transparent)]
+    Chainstate(#[from] chainstate::ChainstateError),
+    #[error("Could not find the previous tip")]
+    OldTip,
+    #[error("Could not find the previous tip index")]
+    OldTipIndex,
+    #[error("Could not find the new tip index")]
+    NewTipIndex,
+    #[error("Block {0:?} not found while traversing history")]
+    BlockNotFound(Id<Block>),
+    #[error("Chainstate call: {0}")]
+    ChainstateCall(#[from] subsystem::subsystem::CallError),
+}
+
+/// Collect blocks between the given two points
+fn collect_blocks<C: ChainstateInterface + ?Sized>(
+    chainstate: &C,
+    mut curr_id: Id<GenBlock>,
+    stop_id: Id<GenBlock>,
+) -> Result<Vec<Block>, ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    let mut result = Vec::new();
+    while curr_id != stop_id {
+        let curr_block_id = curr_id
+            .classify(chain_config)
+            .chain_block_id()
+            .expect("Reached genesis before the stopping block");
+        let block = chainstate
+            .get_block(curr_block_id)?
+            .ok_or_else(|| ReorgError::BlockNotFound(curr_block_id))?;
+        curr_id = block.prev_block_id();
+        result.push(block);
+    }
+    Ok(result)
+}
+
+/// Blocks affected by a reorg
+struct ReorgInfo {
+    disconnected: Vec<Block>,
+    connected: Vec<Block>,
```

Comment:
Please add a comment about the order of these blocks. It seems that both are in reverse chronological order (higher height first)